pkg/queue/concurrency_state_test.go

```diff
@@ -17,116 +17,127 @@ limitations under the License.
-				})
-			}
+	if got, want := pollFor(resumed, 2), int64(2); got != want {
+		t.Errorf("Resume was called %d times, want %d times", got, want)
+	}
+}
 
-			wg.Wait()
-			// Allow last update to finish (otherwise values are off, though this doesn't show
-			// as a race condition when running `go test -race `
-			// TODO Less hacky fix for this
-			time.Sleep(100 * time.Microsecond)
+func TestConcurrencyStateHandlerParallelSubsumed(t *testing.T) {
+	paused := atomic.NewInt64(0)
+	resumed := atomic.NewInt64(0)
 
-			if got, want := paused.Load(), tt.pauses; got != want {
-				t.Errorf("expected to be paused %d times, but was paused %d times", want, got)
-			}
+	req1 := make(chan struct{})
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Query().Get("req") == "1" {
+			req1 <- struct{}{} // to know it's here.
+			req1 <- struct{}{} // to make it wait.
+		}
+	}
+	logger := ltesting.TestLogger(t)
+	h := ConcurrencyStateHandler(logger, http.HandlerFunc(handler), func() { paused.Inc() }, func() { resumed.Inc() })
 
-			if got, want := delegated.Load(), int64(len(tt.events)); got != want {
-				t.Errorf("expected to be delegated %d times, but delegated %d times", want, got)
-			}
+	go func() {
+		defer func() { req1 <- struct{}{} }()
+		h.ServeHTTP(httptest.NewRecorder(), httptest.NewRequest("GET", fmt.Sprintf("http://target?req=%d", 1), nil))
+	}()
 
-			if got, want := resumed.Load(), tt.resumes; got != want {
-				t.Errorf("expected to be resumed %d times, but was resumed %d times", want, got)
-			}
-		})
+	<-req1 // Wait for req1 to arrive.
+
+	// Send a second request, which can pass through.
+	h.ServeHTTP(httptest.NewRecorder(), httptest.NewRequest("GET", "http://target", nil))
+
+	<-req1 // Allow req1 to pass.
+	<-req1 // Wait for req1 to finish.
+
+	if got, want := pollFor(paused, 1), int64(1); got != want {
+		t.Errorf("Pause was called %d times, want %d times", got, want)
+	}
+
+	if got, want := pollFor(resumed, 1), int64(1); got != want {
+		t.Errorf("Resume was called %d times, want %d times", got, want)
+	}
+}
+
+func TestConcurrencyStateHandlerParallelOverlapping(t *testing.T) {
+	paused := atomic.NewInt64(0)
+	resumed := atomic.NewInt64(0)
+
+	req1 := make(chan struct{})
+	req2 := make(chan struct{})
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Query().Get("req") == "1" {
+			req1 <- struct{}{} // to know it's here.
+			req1 <- struct{}{} // to make it wait.
+		} else {
+			req2 <- struct{}{} // to know it's here.
+			req2 <- struct{}{} // to make it wait.
+		}
+	}
+	logger := ltesting.TestLogger(t)
+	h := ConcurrencyStateHandler(logger, http.HandlerFunc(handler), func() { paused.Inc() }, func() { resumed.Inc() })
+
+	go func() {
+		defer func() { req1 <- struct{}{} }()
+		h.ServeHTTP(httptest.NewRecorder(), httptest.NewRequest("GET", fmt.Sprintf("http://target?req=%d", 1), nil))
+	}()
+
+	<-req1 // Wait for req1 to arrive.
+
+	go func() {
+		defer func() { req2 <- struct{}{} }()
```

Comment:
not super important, but out of interest I'm wondering why this needs to be a defer, are we expecting that ServeHTTP could panic here?

Reply:
yep, just "best-practice" whenever something is handed off via ServeHTTP, we need to expect it to panic. In this case, we're sort of guaranteed it won't but it felt like it doesn't hurt either.

Reply:
fair enough